File: FindCycle.java
import java.util.HashSet;

public class FindCycle {
	
	//Brute Force Approach which takes O(n**2) time
	public static int getCycle(int i, int[]nums) {
		
		HashSet<Integer> set=new HashSet<Integer>();
		int t = nums[i];
		
		while(!set.contains(t)) {
			set.add(t);
			t=nums[t];
		}
		
		return set.size();
	}
	
	public static int biggestCycle(int[] nums) {
		
		int longestCycle=0;
		
		for (int i=0; i<nums.length;i++) {
			
			int t = getCycle(i, nums);
			if( t > longestCycle) {
				longestCycle = t;
			}
		}
		return longestCycle;
	}
	
	
	
	
	public static void main(String[] args) {
		
		int[] numbers = {0,1,2};
		
		int q = biggestCycle(numbers);
		System.out.println("Der längste Zyklus ist: "+q);

	}

}

File: FindCycle.py
import numpy as np
from numpy.lib import math
# Given an integer array nums of length n where nums is a permutation of the numbers in the range [0..n-1]
# You should build a set s[k]= {nums[k],nums[nums[k]],...} subjected to the following rule:
# 1)The first element in s[k] starts with the section of the element nums[k] of index = k
# 2)The next element in s[k] should be nums[nums[k]] and then nums[nums[nums[k]]] and so on
# 3)We stop adding right before a duplicate element occurs in s[k]

# Return the length of the longest set s[k]

# Approach One: Brute force


def getCycle(i: int, nums: list):
    set = {}
    t = nums[i]

    while(t not in set):
        set.add(t)
        t = nums[t]

    return len(set)


def longestCycle(nums: list):
    longestCyc = 0
    for i in nums:
        t = getCycle(i, nums)
        if(t > longestCyc):
            longestCyc = t

    return longestCyc


firstSet = {5, 4, 0, 3, 1, 6, 3}

print(longestCycle(firstSet))


# Approach 2: Use an array for visited indices

def usingVisitedArray(nums: np.array):
    visited = np.full([len(nums)], False)
    longest = 0

    for x in np.nditer(nums):
        if(not visited[x]):
            start = nums[x]
            count = 0

            while (start != nums[x]):
                start = nums[start]
                count += 1
                visited[start] = True
                res = max(longest, count)

    return longest
